lib/core/src/display/shape/text/text_field/word_occurrence.rs

```diff
@@ -0,0 +1,202 @@
+                    cursor_location.line == current_word.start.line &&
+                        cursor_location.column >= current_word.start.column &&
+                        cursor_location.column <= current_word.end.column
+                }).map(|(index, _)| index).unwrap_or(0);
+                let current_index = current_index - 1;
+                Some(Self { words, current_index }.initialize(&cursor))
+            }
+        } else {
+            None
+        }
+    }
+
+    fn initialize(mut self, cursor:&Cursor) -> Self {
+        if cursor.has_selection() {
+            self.advance()
+        }
+        self
+    }
+
+    fn current_mut(&mut self) -> &mut WordRange {
+        &mut self.words[self.current_index]
+    }
+
+    fn advance(&mut self) {
+        self.current_index = (self.current_index + 1) % self.words.len();
+    }
+
+    /// Get next word occurrence if not already selected.
+    pub fn select_next(&mut self) -> Option<WordRange> {
+        self.advance();
+        let mut word = self.current_mut();
+        if word.is_selected {
+            None
+        } else {
+            word.is_selected = true;
+            Some(word.clone())
+        }
+    }
+}
+
+
+
+// =============
+// === Utils ===
+// =============
+
+fn get_words(content:&[char]) -> Vec<Vec<(usize,char)>> {
+    let indexed:Vec<(usize,char)> = content.iter().copied().enumerate().collect();
+    let words = indexed.split(|(_,character)| !character.is_alphanumeric() && *character != '_');
+    let words = words.filter(|word_in_context| !word_in_context.is_empty());
+    words.map(|c| c.to_vec()).collect()
+}
+
+fn get_index_range_of_word_at(content:&[char], index:usize) -> Option<Range<usize>> {
+    let words       = get_words(content);
+    let mut ranges  = words.iter().map(|word| {
+        let (start,_) = word.first().unwrap();
+        let end       = start + word.len();
+        *start..end
+    });
+    ranges.find(|word_range| index >= word_range.start && index <= word_range.end)
+}
+
+fn get_word_occurrences(content:&[char], word:&[char]) -> Vec<Range<usize>> {
+    let mut occurrences = Vec::new();
+
+    for word_in_content in get_words(content) {
+        let count = word_in_content.iter().zip(word).filter(|&((_, a), b)| a == b).count();
+        if count == word_in_content.len() {
+            let (start,_) = word_in_content.first().unwrap();
+            let end       = start + word_in_content.len();
+            occurrences.push(*start..end)
+        }
+    }
+
+    occurrences
+}
+
+
+
+// =============
+// === Tests ===
+// =============
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn index_range_of_word_at() {
+        let content           = String::from("   abc    def    ghi  ");
+        let content:Vec<char> = content.chars().collect();
+        let range             = get_index_range_of_word_at(&content,12);
+        assert_eq!(range, Some(10..13));
+    }
```

Comment:
Please add cases for string begin and end, as well as for empty string. And the _ and numbers should be also covered.